System/Directory/Template.hs

```diff
@@ -0,0 +1,1561 @@
+-- Implementation-dependent constraints are not satisfied.
+-- @[EBUSY]@
+--
+-- * @UnsupportedOperation@
+-- The implementation does not support renaming in this situation.
+-- @[EXDEV]@
+--
+-- * @InappropriateType@
+-- Either the destination path refers to an existing directory, or one of the
+-- parent segments in the destination path is not a directory.
+-- @[ENOTDIR, EISDIR, EINVAL, EEXIST, ENOTEMPTY]@
+--
+-- @since 1.2.7.0
+renamePath :: FILEPATH                  -- ^ Old path
+           -> FILEPATH                  -- ^ New path
+           -> IO ()
+renamePath opath npath =
+  (`ioeAddLocation` "renamePath") `modifyIOError` do
+    renamePathInternal opath npath
+
+-- | Copy a file with its permissions.  If the destination file already exists,
+-- it is replaced atomically.  Neither path may refer to an existing
+-- directory.  No exceptions are thrown if the permissions could not be
+-- copied.
+copyFile :: FILEPATH                    -- ^ Source filename
+         -> FILEPATH                    -- ^ Destination filename
+         -> IO ()
+copyFile fromFPath toFPath =
+  (`ioeAddLocation` "copyFile") `modifyIOError` do
+    atomicCopyFileContents fromFPath toFPath
+      (ignoreIOExceptions . copyPermissions fromFPath)
+
+-- | Copy the contents of a source file to a destination file, replacing the
+-- destination file atomically via @withReplacementFile@, resetting the
+-- attributes of the destination file to the defaults.
+atomicCopyFileContents :: FILEPATH            -- ^ Source filename
+                       -> FILEPATH            -- ^ Destination filename
+                       -> (FILEPATH -> IO ()) -- ^ Post-action
+                       -> IO ()
+atomicCopyFileContents fromFPath toFPath postAction =
+  (`ioeAddLocation` "atomicCopyFileContents") `modifyIOError` do
+    withReplacementFile toFPath postAction $ \ hTo -> do
+      copyFileToHandle fromFPath hTo
+
+-- | A helper function useful for replacing files in an atomic manner.  The
+-- function creates a temporary file in the directory of the destination file,
+-- opens it, performs the main action with its handle, closes it, performs the
+-- post-action with its path, and finally replaces the destination file with
+-- the temporary file.  If an error occurs during any step of this process,
+-- the temporary file is removed and the destination file remains untouched.
+withReplacementFile :: FILEPATH            -- ^ Destination file
+                    -> (FILEPATH -> IO ()) -- ^ Post-action
+                    -> (Handle -> IO a)    -- ^ Main action
+                    -> IO a
+withReplacementFile path postAction action =
+  (`ioeAddLocation` "withReplacementFile") `modifyIOError` do
+    mask $ \ restore -> do
+      -- TODO: AFPP doesn't support openBinaryTempFile yet,
+      --       so we have to use this (sad) workaround
+      --       (on unix, converts using filesystem encoding, on windows
+      --       converts with UTF-16LE)
```

Comment:
This is a much bigger wart... it means we'll have to inline some code from base:

* https://gitlab.haskell.org/ghc/ghc/-/blob/67072c31d8b6ce4f0de79fa52bc3e5cdd5a495c6/libraries/base/System/IO.hs#L501-591

@Mistuke said 

> <Phyx-> maerwald: you can inline a simplified version, the one from handleResultsWinIO, findTempNamePosix is legacy but I hadn't removed it yet as we'll remove it all when winio becomes the only one. but difference is only in what they do with the filePath after it's returned. For that you can call your new CreateFile with AFPP support and then call Win32's hANDLEToHandle again

Reply:
Can this be implemented in the `file-io` library rather than here?

Reply:
No, because it's not a boot library.